In [1]:
library(dplyr)

Warning message:
"package 'dplyr' was built under R version 4.0.5"

Attaching package: 'dplyr'


The following objects are masked from 'package:stats':

    filter, lag


The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union




**Regresión Poisson:**

Variable respuesta de conteo o bien expresada en relación-tasa

***Ejercicio:*** El número de premios que recibe un estudiante a lo largo de su bachillerato, según el programa de estudio y calificación en matemáticas.

In [11]:
df=read.csv("https://stats.idre.ucla.edu/stat/data/poisson_sim.csv")
head(df,10)

id  num_awards prog math
1   45 0          3    41  
2  108 0          1    41  
3   15 0          3    44  
4   67 0          3    42  
5  153 0          3    40  
6   51 0          1    42  
7  164 0          3    46  
8  133 0          3    40  
9    2 0          3    33  
10  53 0          3    46

**Tarea:** Hacer las gráficas necesarias y un pequeño dashboard que les sea interesante.

In [12]:
str(df)

'data.frame':	200 obs. of  4 variables:
 $ id        : int  45 108 15 67 153 51 164 133 2 53 ...
 $ num_awards: int  0 0 0 0 0 0 0 0 0 0 ...
 $ prog      : int  3 1 3 3 3 1 3 3 3 3 ...
 $ math      : int  41 41 44 42 40 42 46 40 33 46 ...


In [5]:
#Variable Prog: {1: 'General',2: 'Academic', 3:'Vocational'}

df$prog=as.factor(df$prog)
str(df)

'data.frame':	200 obs. of  4 variables:
 $ id        : int  45 108 15 67 153 51 164 133 2 53 ...
 $ num_awards: int  0 0 0 0 0 0 0 0 0 0 ...
 $ prog      : Factor w/ 3 levels "1","2","3": 3 1 3 3 3 1 3 3 3 3 ...
 $ math      : int  41 41 44 42 40 42 46 40 33 46 ...


In [ ]:
#¿Qué harían para retiquetar las variables de prog? Es decir, 
# en lugar de número, que muestren la etiqueta (texto)

#Tarea: Cambiar los números por las etiquetas 

if(pr)
    
verygood <- ifelse(tablaD$cut=='Very Good',1,0)




In [13]:
#Cambiar a númerico
df$new_prog<-cut(df$prog, breaks = c(-Inf,1,2,Inf), labels = c("General","Academic","Vocational"))

In [14]:
str(df)

'data.frame':	200 obs. of  5 variables:
 $ id        : int  45 108 15 67 153 51 164 133 2 53 ...
 $ num_awards: int  0 0 0 0 0 0 0 0 0 0 ...
 $ prog      : int  3 1 3 3 3 1 3 3 3 3 ...
 $ math      : int  41 41 44 42 40 42 46 40 33 46 ...
 $ new_prog  : Factor w/ 3 levels "General","Academic",..: 3 1 3 3 3 1 3 3 3 3 ...


In [7]:
df$prog <- ifelse(df$prog==1,'General', df$prog)
df$prog <- ifelse(df$prog==2,'Academic', df$prog)
df$prog <- ifelse(df$prog==3,'Vocational', df$prog)

In [15]:
df$prog=as.factor(df$prog)

In [16]:
str(df)

'data.frame':	200 obs. of  5 variables:
 $ id        : int  45 108 15 67 153 51 164 133 2 53 ...
 $ num_awards: int  0 0 0 0 0 0 0 0 0 0 ...
 $ prog      : Factor w/ 3 levels "1","2","3": 3 1 3 3 3 1 3 3 3 3 ...
 $ math      : int  41 41 44 42 40 42 46 40 33 46 ...
 $ new_prog  : Factor w/ 3 levels "General","Academic",..: 3 1 3 3 3 1 3 3 3 3 ...


In [18]:
levels(df$prog)[levels(df$prog)=='1']<-'General'

In [19]:
levels(df$prog)

[1] "General" "2"       "3"

In [20]:
levels(df$prog)[levels(df$prog) %in% '2']<-'Academic'
levels(df$prog)[levels(df$prog) %in% '3']<-'Vocational'

In [21]:
levels(df$prog)

[1] "General"    "Academic"   "Vocational"

In [ ]:
#library(purrr) library(cats) universo tidyverse
#library (recipes) wrangling data 'featuring engineering'....que en español limpieza

In [23]:
#Modelo de regresión Poisson
poi1<-glm(num_awards~prog+math,data=df,family='poisson')
summary(poi1)
#variable pivote para prog es 'General', si queremos cambiar entonces relevel


Call:
glm(formula = num_awards ~ prog + math, family = "poisson", data = df)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.2043  -0.8436  -0.5106   0.2558   2.6796  

Coefficients:
               Estimate Std. Error z value Pr(>|z|)    
(Intercept)    -5.24712    0.65845  -7.969 1.60e-15 ***
progAcademic    1.08386    0.35825   3.025  0.00248 ** 
progVocational  0.36981    0.44107   0.838  0.40179    
math            0.07015    0.01060   6.619 3.63e-11 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for poisson family taken to be 1)

    Null deviance: 287.67  on 199  degrees of freedom
Residual deviance: 189.45  on 196  degrees of freedom
AIC: 373.5

Number of Fisher Scoring iterations: 6


In [25]:
ls(summary(poi1))

[1] "aic"            "aliased"        "call"           "coefficients"  
 [5] "contrasts"      "cov.scaled"     "cov.unscaled"   "deviance"      
 [9] "deviance.resid" "df"             "df.null"        "df.residual"   
[13] "dispersion"     "family"         "iter"           "null.deviance" 
[17] "terms"

In [27]:
exp(summary(poi1)$coefficients) #odds ratio

Estimate   Std. Error z value      Pr(>|z|)
(Intercept)    0.00526263 1.931802   3.460712e-04 1.000000
progAcademic   2.95606545 1.430828   2.060228e+01 1.002486
progVocational 1.44745846 1.554370   2.312747e+00 1.494491
math           1.07267164 1.010656   7.489312e+02 1.000000

Lo anterior se interpreta como tasa de incicencia

prog='Academic' exp(B) es igual a 2.96, la tasa de incidencia de prog='Academic' es 2.96 veces mayor que 'General' (la pivote), si el resto de variables permanece constante.

Respecto a math, el número de premios aumenta 7% por cada incremento de unidad en math

Tarea: hacer los pronósticos